### ISSUE TITLE
Investigate showing the full path to base folder

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#9809]
While setting up the "Signal" messenger app I stumbled upon the fact, that it is somehow possible to show the absolute path to a folder selected via SAF.
I set up the backup directory in that app via SAF and it shows the complete path now in the settings. In my case: `Lars 512GB/AppBackup/Signal Messenger`. 

So it seems it is somehow possible for an app to know the absolute path (or did I overlook some requirement for that, which c:geo currently does not have)?
It would IMHO be very helpful for our users to see that as well instead of only the prefixed `...`.

**Comment by eddiemuc** [comment]
@Lineflyer Thanks a lot, finally I know of an app using Scoped Storage to check what is possible against it!

I installed Signal and played around with selecting different folders on my device seeing what it will display. In my case (OnePlus 7T) however it seems to use same kind of extraction mechanism that I try to use too:
* If I select a user-created "free folder" (e.g. `/myfolders/Signal`), it shows the "path" part without the "`/tree/primary:`" before it - same that I currently do for c:geo except I add a "`.../`" before it
* Specifically I tested a use case which puzzled me for c:geo for a while: if I select a subfolder under "`Documents`" (e.g. /Documents/Signal" it just shows me the subfolder name (e.g "`Signal`"), leaving out "`Documents`". (@Lineflyer it would be interesting to know if the same happens for you). This is the behaviour I had to deal with for c:geo too. I found out that in case of selecting "Documents", the Uri is prepended with "`/tree/home:`", and for c:geo I prepended "`Documents`" to the path in this case

You may observe and comment my "Uri-for-user-format-logic" in `UriUtils.toUserDisplayableString`, and you may see some of the more strange Uri examples I had to deal with in class `UriUtilsTest`. Note that the path path is encoded and has to be decoded first. Examples are:

```
    //Example for folder /cgeo
    private static final String DOC_URI_EXAMPLE = "content://com.android.externalstorage.documents/tree/primary%3ADocuments%2Fcgeo/document/primary%3ADocuments%2Fcgeo%2Flogfiles%2Flogcat_2020-12-28_17-22-20-2.txt";
    //Example for folder /Documents/cgeo
    private static final String DOC_URI_EXAMPLE_2 = "content://com.android.externalstorage.documents/tree/home%3Acgeo";
    //Example for folder /Downloads/cgeo
    private static final String DOC_URI_EXAMPLE_3 = "content://com.android.externalstorage.documents/tree/primary%3ADownload%2Fcgeo";

    //complicated example for a document in /Documents/cgeo/gpx
    private static final String DOC_URI_COMPLICATED = "content://com.android.externalstorage.documents/tree/home%3Acgeo/document/home%3Acgeo%2Fgpx%2Froute_2021-01-06_17-14-08-4.gpx";
```

Uris may be different on different devices or SDKs or whatever of course, I simply don't know. The above examples are from the Android Studio emulators for SDK29 and SDK30.

So, in conclusion: I would be really glad for extended knowledge here, I surely don't know everything about this. But from what I currently saw in Signal, they try the best to extract user-readable paths out of Uris same as we do.

**Comment by fm-sys** [comment]
Maybe we can take a look at there source code, as they are also here on github. See [signalapp/signal-android](https://github.com/signalapp/Signal-Android)

**Comment by eddiemuc** [comment]
Looking into the SignalApp code, the following is their method to get a user-readable path from an Uri. Note that it requires API24:

https://github.com/signalapp/Signal-Android/blob/7e64d57ba8fc8ac295c7c35279fba2bf6400655b/app/src/main/java/org/thoughtcrime/securesms/util/StorageUtil.java#L61-L83

For reference, the current value of `StorageUtil__s_s` is:

https://github.com/signalapp/Signal-Android/blob/7e64d57ba8fc8ac295c7c35279fba2bf6400655b/app/src/main/res/values/strings.xml#L3005-L3006

**Comment by Lineflyer** [comment]
> Note that it requires API24

Does this mean we could implement it but restrict to API24+ or does it mean we can only use it once we do not support API<24 any more?

>  (@Lineflyer it would be interesting to know if the same happens for you).

Still interesting or maybe already answered by the fact, that the device you used is below API 24?

**Comment by Lineflyer** [comment]
> (@Lineflyer it would be interesting to know if the same happens for you).

Tested it anyway:
It turned out it only happens to be complete if I use a folder below my SD-card

If I use a subfolder of my internal root (e.g. `/a/b`) it shows `a/b`. Same applies also if I use `/Documents` as directory `a`
So at least it still shows a readable path starting from `/` in case I use the internal memory instead of my SD-card

Info: Samsung S20, Android 11

**Comment by moving-bits** [comment]
> > Note that it requires API24
> Does this mean we could implement it but restrict to API24+ or does it mean we can only use it once we do not support API<24 any more?

We could use this function on systems with API > 23, and stay with "..." on older systems.

**Comment by eddiemuc** [comment]
Hi @Lineflyer and @moving-bits ,

what I take away from this is that there is no "natural" way of displaying content Uris to users, and Signal struggled and created their own interpreting method for this, same as we did. We can take a look at their method and take over the aspects we like of course (from what I see they try to display a volume name too, which I currently didn't consider). But their method is not perfect as well (e.g. they don't handle the Document folder case)

Wrt api24: I did not further analyze yet which parts of the function use API24, maybe it is the StorageManager they use for getting volume names. In any case, I am sure it is possible to find a way handling the API<24 case gracefully.

What we need to do is figure out our own c:geo way of displaying Uris to users, and the Signal method can give us hints as to what is possible and what not.

**Comment by eddiemuc** [comment]
> > (@Lineflyer it would be interesting to know if the same happens for you).
> 
> Tested it anyway:
> It turned out it only happens to be complete if I use a folder below my SD-card
> 
> If I use a subfolder of my internal root (e.g. `/a/b`) it shows `a/b`. Same applies also if I use `/Documents` as directory `a`
> So at least it still shows a readable path starting from `/` in case I use the internal memory instead of my SD-card
> 
> Info: Samsung S20, Android 11

Strange, it is different on my device. For `/Documents/Signal/abc` I get folder `Signal/abc` (without leading /). For `/Signal/abc` I get `Signal/abc` as well (also without leading /). I have no sd card installed on my device.

**Comment by Lineflyer** [comment]
> Strange, it is different on my device. For /Documents/Signal/abc I get folder Signal/abc (without leading /). For /Signal/abc I get Signal/abc as well (also without leading /). I have no sd card installed on my device.

The leading `/` is also missing for me as described above.
But for the folders below `/Documents` I also get the same structure, e.g. `Documents/test`
Is `Dcouments` something special? The folder was already there on my device (and I did not create it myself), but I never used it AFAIK.

**Comment by eddiemuc** [comment]
> Is `Dcouments` something special? The folder was already there on my device (and I did not create it myself), but I never used it AFAIK.

Documents is special in the same way as folders Pictures, DCIM, Music, Downloads, Videos etc in that it is a folder associated with a "MediaStoreAPI" collection, namely the "Files" collection (the collection of all "non-media-files" within the MediaStore). As far as I investigated it seems that both "Download" and "Documents" are associated with this collection (it is not unique that there are multiple folders associated with one collection, e.g. "DCIM" and "Pictures" are both associated with the Image collection)

The special handling of document Uris inside the Documents folder I discovered by code debugging on the emulators. I did not find documentation about this (documentation basically says that you can never rely on Uri formats for anything which is whay this is all a big trial & error in my understanding). Strange that this is not the case on your device.

@Lineflyer you could do me a real favor if you install the SAF version on your device, set a folder (maybe the offline map folder) to a subdirectory of "Documents" (on the internal storage, not SD card) and then send me the system information (it will include the untampered raw Uri which is of interest to me). Sorry to bother you, you seem to be the only one trying out this stuff currently.

**Comment by eddiemuc** [comment]
Another funny artifact when using "Signal" (which however is consistent with my own experience with SAF): If I choose a folder I can select as different "root sources" either my device ("OnePlus 7T") or "Downloads". If I choose "Downloads" then I am brought to the "Download" folder (no surprise here). If I now create a subfolder there (e.g. "Signal") and choose this folder, the "folder path" I get is just a number (1793 in my case).

If I choose the very same folder starting with "root source" OnePlus 7T then I get the expected path.

The solution to this behaviour is: When I choose the "root source" "Downloads", what happens behind the scenes is that Android uses a different document provider, namely one using the MediaStoreAPI. And this API uses a different Uri sceme, namely assigning unique-id uris to all elements. Hence the strange "number". 

So again: there is no "natural way" to display a content uri in Android, and we simply have to find our own compromise.

**Comment by eddiemuc** [comment]
Screenshots for last comment:

![Screenshot_20210118-103422.jpg](https://user-images.githubusercontent.com/6909759/104897802-1b5e1e80-5979-11eb-8859-5b19aa397143.jpg)

![Screenshot_20210118-103450.jpg](https://user-images.githubusercontent.com/6909759/104897811-2022d280-5979-11eb-92f2-b323293bd70f.jpg)

![Screenshot_20210118-103459.jpg](https://user-images.githubusercontent.com/6909759/104897827-244ef000-5979-11eb-917e-d77bd995da7e.jpg)

I am pretty sure the same would happen for c:geo with SAF

**Comment by fm-sys** [comment]
I'm using signal since many years and had never seen such a dialog forcing me to grant access to a folder for SAF (but automatic backup to file system is also still activated).

I guess that they are using the deprecated API with the migration flag enabled until the app gets uninstalled.

To be honest, I am a bit afraid that our upgrade dialog will confuse many users as it is quite technical. 

Therefore we should IMHO think about:
- Show the dialog only when the folder can no longer be accessed by c:geo elsewhere 
- don't provide the options "copy, move, do nothing" as probably users don't know what to choose best. The easier the better. That is why I would like to plead for automatically identifying the best method for the user. Are c:geo files in the folder selected via SAF contained (it is the same folder as before) we don't need to move anything. If another folder is selected, the files must be moved before c:geo can work properly. Copying makes not really sense IMO, because all the data (especially map files that require a lot of disk space) is available twice afterwards, which inevitably leads to confusion.

**Comment by eddiemuc** [comment]
What @fm-sys suggests is absolutely possible with little changes to the implementation.

Some things to consider:
* We can make the dialog pop up only if the legacy dirs can't be accessed any more. Naturally, of course, then we can't copy or move the content of this dir to the new dir and user has to do that via e.g. file explorer app (or start with fresh dirs)
* We can leave out choices for copy/move and pick what we think is best for the user. Then however we are responsible that this is really the best in all cases. This gives a lot responsiblities to our testers, to make sure the selected logic really serves the needs of all users on all devices

**Comment by fm-sys** [comment]
> We can make the dialog pop up only if the legacy dirs can't be accessed any more. Naturally, of course, then we can't copy or move the content of this dir to the new dir and user has to do that via e.g. file explorer app (or start with fresh dirs)

That's indeed a valid point. Has Google given any information as to when this will be the case?

I think when leave out choices for copy/move it should probably be quite self-explanatory to configure  SAF for end-users.

Now I'm just worried about developer/testers. Personally, I switch very often between self-signed, nightly and PR apk's and therefore always need to uninsall c:geo. Currently I skip the installation wizard and click restore backup in the dialog appearing in the Main Activity. That doesn't cost me more than 5 seconds. Unfortunately, this will not be that easy with SAF...

**Comment by Lineflyer** [comment]
> That's indeed a valid point. Has Google given any information as to when this will be the case?

Not explicitly AFAIK. 
But it will happen as soon as they enforce to target SDK30 IMHO. 
This *could* be in November 2021 assuming its one year after SDK29 enforcement.

**Comment by eddiemuc** [comment]
> > That's indeed a valid point. Has Google given any information as to when this will be the case?
> 
> Not explicitly AFAIK.
> But it will happen as soon as they enforce to target SDK30 IMHO.
> This _could_ be in November 2021 assuming its one year after SDK29 enforcement.

In addition to what @Lineflyer said:
when switching to SDK30 you can set a flag "preserveLegacyExternalStorage" in the manifest. This is meant to support migration and leads to legacy dirs still being accessible, but **only for apps already installed before the upgrade to SDK30** and until they are deinstalled (I don't know how long the flag still allows access to legacy dirs when the app is further upgraded to new releases). This scenario might be helpful but I think it is really hard to test it.

Having testers situation in mind, I specifically allowed to set and reset the base directory of c:geo also back to the legacy dir. This way the migration situation can be replayed as often as wanted. Also, to allow for intensitve testing and real-life usage before  everything it is too late, my proposal is to roll-out the SAF change in a first release with SDK29 still turned on. That way we are able to react to problems discovered in the field without having all possible paths blocked. The change is really very big, and I would appreciate it if we keep this chance.

**Comment by eddiemuc** [comment]
See PR #9912

**Comment by MagpieFourtyTwo** [comment]
Sorry for writing in this closed issue, but perhaps I simply missed something discussed in another thread, thus don't want to bring trouble into probably already cleared stuff ...

However: Just stumbled upon this thread while tracing [commit 5c1e56d](https://github.com/cgeo/cgeo/commit/5c1e56ddc3a139ded286cd1b6813239e7a6e8cad) - cause although I expected to see clearer strings for folder names in the settings now, I still just get ``/cgeo (User defined)`` as value for ``Base folder for public local data`` in ``Settings`` => ``System``, which IMO is kind of confusing, cause it completely misses any location information at all.  Same for ``Offline Data`` => ``GPX`` => ``GPX Folder``, which just displays as ``/Download (User-defined)``. Even worse, it gives the impression this stuff will go to Internal Memory ... but for real they are on Internal SD card.

My "real" values are:
- Base folder: ``DOCUMENT::content://com.android.externalstorage.documents/tree/primary%3Acgeo::``
- GPX folder: ``DOCUMENT::content://com.android.externalstorage.documents/tree/primary%3ADownload::``

Had in fact not enough time to check the implementation completely, but got the impression, it has been implemented in this way on purpose, although the (IMO important) info if this folder resides in Internal Memory, Internal SD card or even on External SD card got lost completely - and should be added in some way,

Moreover it would even be clearer to distinguish between "Internal SD card" and "External SD card", too ...?

**Comment by fm-sys** [comment]
Whats the difference between "Internal Memory" and "Internal SD card"? Is "Internal SD card" a real "SD card" and what is meant with "External SD card"? I would very much like to help finding a solution, but I do not understand the three terms... 

Possible solution (pure guess): Can "primary" always be decoded as "/emulated/0/"? The first thing we need to do is determine what the url should look like and see how it can be derived from the tree-uri.

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo no problem with reopening this issue. Indeed current implementation would display Uri to user just as you described it. 

If I understand you would like to prepend something like "External storage" if the uri provider is "com.android.externalstorage.documents"? Would be easy to implement.

I have a favor to ask though: could you select a folder residing on your **internal** storage and send me the uri you get? Because in my tests I always got Uris with the "externalstorage" provider when selecting a folder on internal storage (which puzzled me), and I got Uris with a uuid in it when selecting from sd card. This is why I thought (maybe wrongly) that despite the strange name the above Uri format would refer to internal storage.

**Comment by MagpieFourtyTwo** [comment]
Don't know how much is really known already, so perhaps I'm bit of captain obvious right now, but the questions above made me a bit uncertain ... so perhaps better more than too less. ;)

However, storage terminology is indeed a bit weird, cause it seems to defy logic. But seen from system's perspective, it may make perfect sense: If you see Android's system memory as "Internal" (as it usually is hidden from users and accessible on rooted devices only, anyway), then the memory for user data is most obviously "External". Thus generally *all* memory a user can access on "official" way can be called "externalstorage". However, there are two different external storages: Built-In (so-called) "SD cards" (which are in fact just some more non-removable internal memory, just a bit slower [read "cheaper"] than system's internal memory), and OTOH there are removable "real" SD cards, which then are "Ext SD cards". Thus the latter is de facto an external externalstorage ...  ;)

Over here things look like:

storage | real path | persistable | currently shown as | poss. better shown as
--------|----------------|-----------|----------|-
External SD card | /storage/0123-4567/cgeo | DOCUMENT::content: //com.android.externalstorage.documents/tree/0123-4567%3Acgeo:: | SD card/cgeo | External SD card/cgeo
"" | /mnt/ext_sdcard/cgeo | DOCUMENT::content: //com.android.externalstorage.documents/tree/0123-4567%3Acgeo:: | SD card/cgeo | External SD card/cgeo
Internal SD card | /sdcard/cgeo | DOCUMENT::content: //com.android.externalstorage.documents/tree/primary%3Acgeo:: | /cgeo | Internal SD card/cgeo
"" | /storage/emulated/0/cgeo | DOCUMENT::content: //com.android.externalstorage.documents/tree/primary%3Acgeo:: | /cgeo | Internal SD card/cgeo
"" | /mnt/sdcard/cgeo | DOCUMENT::content: //com.android.externalstorage.documents/tree/primary%3Acgeo:: | /cgeo | Internal SD card/cgeo
Internal Memory | /cgeo | no example, as current device is not rooted | /cgeo | /cgeo

Hope this is understandable ...

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo  Thanks a lot for the exaustive analysis!

I understand the request. SOme additional context might be helpful:
* the Android system contains a list of Storage Volumnes which can be accessed by c:geo. Among other things it contains for each volume (identified by it's UUID) a volume name. In the first two rows above, the volumne name "SD card" is retrieved this way using the volume UUID 0123-4567 from the Uri. So this is not a c:geo-invented name but comes rather from the system. This logic was copied from the Signal app as suggested by @Lineflyer (and Signal is still the only other app I know which uses scoped storage), and it should work also if the Android system has more than one such external volumes (maybe a theoretical case... are there smartphones with more than one external sd card?)
* The Android system also provides a name for the internal storage. In the emulators this value is usually set to "internal shared storage".  Currently it is not displayed.

To stay close to the Android provided information and not inventing own names, I would suggest to keep displaying the Android volume names for "real" external sd cards and add the display of the Android internal storage name to the Uris pointing to "internal" sd card (or whatever  the device technically uses for internal memory), e.g. "internal shared storage". Would that be ok?

**Comment by eddiemuc** [comment]
Wrt Android storage terminology, one has to understand the concept of "Document Providers" which Android introduced with this whole scoped storage thing in SDK21. The idea is that one can have different "Document Provider" implementations on its system, each being able to handle "content" Uris. The Document Provider implementation to use is identified by the "provider" part of the Uri. In all above examples, the "Document Provider" is "com.android.externalstorage.documents" - a default one which comes with the system. Up to now, I only have faced one other "Document Provider" which is "com.android.providers.downloads.documents". This is a document provider specifically for handling files in the Download folder. Different "Document Providers" may access the same files/folders on the file system, and they may define different Uri structures and identifiers for the same file/folder. 

The idea behind that as far as I understand it is that you can have document providers e.g. also for cloud storage systems or application specific internal storages which by this way can be made available to all apps on the phone using the same system. More info can be found here: https://developer.android.com/guide/topics/providers/document-provider 

Here is an example screenshot from that page where the user can choose between different document providers "Downloads", "SD CARD", "Bug reports" and "Drive":

![image](https://user-images.githubusercontent.com/6909759/106878299-12d84880-66da-11eb-89c2-c747b290c51c.png)

**Comment by Lineflyer** [comment]
@eddiemuc 
Does this mean a user could even select to store his base folder on cloud services (such as Google Drive in your example). That would be a nice feature (and somehow even useful for the stuff c:geo stores there, like backups).

**Comment by Lineflyer** [comment]
> Hope this is understandable ...

Thanks for this table.
Due to the large amount of trials and tests back at the time where @pstorch redesigned the storage (to the state we do now call "legacy" when introducing SAF) I explored the mount points, names, etc. on several devices and Android versions.
While I do not have the detailed results anymore, one thing I remember is, that due to historic reasons and different implementations it looks totally different on one or the other device. It got a little more standardized on devices >=Android 6 or 7.

For example what you call "Internal SD card" is a memory path which is IMHO kind of exotic and you wont find it ony many devices.
So at the end we can try our best to find a good "display name" for the most common URIs but we will never reach a state where it works on all devices. I guess the code derived form the Signal app is rather good (at lest for me it provided the most meaningful display name out of the scoped storage apps I use).

**Comment by eddiemuc** [comment]
@Lineflyer 
> (at lest for me it provided the most meaningful display name out of the scoped storage apps I use

I would be interested in what other apps are using scoped storage already, esp when they are open source, so I can take a peek at their code...

**Comment by eddiemuc** [comment]
> @eddiemuc 
> Does this mean a user could even select to store his base folder on cloud services (such as Google Drive in your example). That would be a nice feature (and somehow even useful for the stuff c:geo stores there, like backups).

In theory yes, but sadly the most (if not all) document providers only support opening of single documents (means: selecting single files) and not of document trees (means: folders). For example, the google drive access document providers does not support doc trees. However, if you pick a single file, e.g. an image from disk for an offline log, you should already see a large list of document providers including e.g. google drive (depends on what you have installed of course).

I found this wrt the topic:
https://stackoverflow.com/questions/28663778/android-open-document-tree-intent-root-locations

**Comment by Lineflyer** [comment]
> I would be interested in what other apps are using scoped storage already, esp when they are open source, so I can take a peek at their code...

I use my favorite filemanager (I guess its not open source)
https://play.google.com/store/apps/details?id=com.alphainventor.filemanager
It AFAICS uses SAF initially to get access to the file system (at least for the external SD) but it does not show any name of it after selecting it....not needed.

Then there is Locus 
https://play.google.com/store/apps/details?id=menion.android.locus
where you will find directory settings in (german) "Einstellungen" - "Verschiedenes" - "Standardverzeichnisse" - "Erweitert" which seems to use SAF under the hood for some of the entries (e.g.  mapsVector and backup).
BTW: Seems Locus manages to make perfect path names out of that.

**Comment by eddiemuc** [comment]
Funny, your favorite file manager is the same as mine. But I have no external sd card, so I can't check this.

Locus Maps on my device seems to use a custom file / dir picker (I use the free version), it looks very different from the Android Document Tree intent found in Signal (and now c:geo :-))

**Comment by eddiemuc** [comment]
I must correct myself. It seems that Locus uses SAF only for the last directory which can be selected (data/strm) and it's own file picker for all others. Maybe they are slowly rolling out SAF too :-)

For this last directory I noticed that when I select a special dir (one which I know behaves a bit different than the others like "Download" via "Download" document provider, or "Documents"), the. Locus simply ignores my selection. Not the best behaviour I would think. But yes, when excluding all "special dirs" it is possible to "guess" the real file name.

**Comment by MagpieFourtyTwo** [comment]
> are there smartphones with more than one external sd card?)

Yes, there are. Remember a mobile designed for developers with two SD card slots and multiboot capability.
A quick search just showed the [Saygus-V2](https://www.phonearena.com/phones/Saygus-V2_id9073) right now, but there were more.

> To stay close to the Android provided information and not inventing own names, I would suggest to keep displaying the Android volume names for "real" external sd cards and add the display of the Android internal storage name to the Uris pointing to "internal" sd card (or whatever the device technically uses for internal memory), e.g. "internal shared storage". Would that be ok?

Sounds good to me.

> Up to now, I only have faced one other "Document Provider" which is "com.android.providers.downloads.documents".

Have seen "[com.android.internalstorage.documents](https://www.javatips.net/api/image-chooser-library-master/image-chooser-library/src/main/java/com/kbeanie/imagechooser/threads/ImageProcessorThread.java)", too - and there seem to be more ... but don't know enough about this topic to say s/th about usability.

> what you call "Internal SD card" is a memory path which is IMHO kind of exotic 

AFAIK all mobiles without a slot for an external SD card have it. At least the ones I could lay my hands on so far.
Nevertheless. it may of course be called differently by different apps ...
BTW: The one I prefer for years now is [cx File Explorer](https://play.google.com/store/apps/details?id=com.cxinventor.file.explorer), although it lost it's multi-tab functionality underway ...
But it allows accessing network storage, too (cloud services as well as network shares, SMB, (S)FTP, WebDAV ...). 
Plus it's able to add folder shortcuts to home screen, which make browsing to frequently accessed folders way more easy.

**Comment by MagpieFourtyTwo** [comment]
OT Addendum: Seems as if [cx File Explorer](https://play.google.com/store/apps/details?id=com.cxinventor.file.explorer) will be retired.  ;)
Was mourning about the lost tab function, thus poked around a bit, and stumbled upon [MiXplorer](https://forum.xda-developers.com/t/mixplorer-q-a-and-faq-user-manual.3308582/#post-65157294) - great stuff.
Esp. with the extensions, downloadable (via app menu) from xda, too. For free (while in GPS it's EUR 4.99 ...)
Although it seems to be a bit unstable ... had to kill it twice already ... but it's definitely worth a try.
